test/Test/GenericHashing.hs

```diff
@@ -0,0 +1,55 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Test.GenericHashing where
+
+import Test.Tasty
+import Test.Tasty.HUnit
+import Foreign hiding (void)
+import Control.Monad (void)
+import Foreign.C
+import LibSodium.Bindings.GenericHashing (cryptoGenericHash, cryptoGenericHashBytes)
+import qualified Data.ByteString.Unsafe as BS
+
+spec :: TestTree
+spec =
+  testGroup
+    "Generic hashing tests"
+    [ testCase "cryptoGenericHash without key" testCryptoGenericHashWithoutKey
+    ,  testCase "cryptoGenericHash with key" testCryptoGenericHashWithKey
+    ]
+
+testCryptoGenericHashWithoutKey :: Assertion
+testCryptoGenericHashWithoutKey =
+  withCStringLen "test test" $ \(cString, cstringLength) ->
+  allocaBytes (fromIntegral cryptoGenericHashBytes) $ \outPtr -> do
+    void $ cryptoGenericHash
+      outPtr
+      cryptoGenericHashBytes
+      (castPtr cString :: Ptr CUChar)
+      (fromIntegral cstringLength)
+      nullPtr
+      0
+    out <- peekCString (castPtr outPtr)
+    assertEqual "Hashed test string is consistent without key"
+                "\DEL=\ETB\SOp\ETBd=\STXBLzL\SOHl2\CAN\DELQ"
+                out
+
+testCryptoGenericHashWithKey :: Assertion
+testCryptoGenericHashWithKey =
+  let key = "af7d1575690407317bf93723a8d1dca5"
+      msg = "Test Test" :: String
+  in
+    withCStringLen msg $ \(cString, cstringLength) ->
+    allocaBytes (fromIntegral cryptoGenericHashBytes) $ \outPtr ->
+    BS.unsafeUseAsCStringLen key $ \(keyPtr, keyLength) -> do
+      void $ cryptoGenericHash
+        outPtr
+        cryptoGenericHashBytes
+        (castPtr cString :: Ptr CUChar)
+        (fromIntegral cstringLength)
+        (castPtr keyPtr)
+        (fromIntegral keyLength)
```

Comment:
Do we have any other way to convert besides `fromIntegral`? I remember there was some apprehension to using it with C type conversion, but I forget where. 

If this is a non-issue, then it's fine.

Reply:
In this case, it really is a non-issue. `CSize` is equivalent to C's `size_t`; the only guarantees are that it's an unsigned whole-number type at least 16 bits wide. The `Int` that you get from `unsafeUseAsCStringLen` is guaranteed at least 22 bits in size, and GHC Haskell ensures it matches a machine integer on the target platform, which is _at least_ 32 bits, if not 64. Arguably there could be a mismatch, but that's not a problem we can prevent, since it's _entirely_ C-side-dependent.